lib/tesla/middleware/retry.ex

```diff
@@ -42,7 +50,9 @@ defmodule Tesla.Middleware.Retry do
   - `:delay` - The base delay in milliseconds (positive integer, defaults to 50)
   - `:max_retries` - maximum number of retries (non-negative integer, defaults to 5)
   - `:max_delay` - maximum delay in milliseconds (positive integer, defaults to 5000)
-  - `:should_retry` - function to determine if request should be retried
+  - `:should_retry` - function with an arity of 1 or 3 used to determine if the request should
+      be retried the first argument is the result, the second is the env and the third is
+      the context: options + `:retries` (defaults to a match on `{:error, _reason}`)
```

Comment:
the best way to document such complexity is by having a `@type` maybe called `should_retry_func` and document it there.

You do not have to do it, just commeting so that in the future others could find this comment.

Reply:
I wanted to do something like that, but I haven't seen anything of the sort in any of the other middlewares... If you want, I can make a follow-up PR that documents all the options. The question is should I leverage typedocs and remove the `## Options` section in favor of using something like:

```elixir
@typedoc "Supported options."
@type options :: %{optional(:delay) => delay(), ...}

@typedoc "The base delay in milliseconds (positive integer, defaults to 50)"
@type delay :: pos_integer()
```

And then just reference it with `t:options/0` in the moduledoc?

Reply:
I tend to say yes, lets give it a try